<a href="https://colab.research.google.com/github/lcalderon-aceituno/PS353-SP22-HW/blob/main/HW5_6_23_Advanced_Dynamics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 6.23


In [95]:
from sympy import *
from sympy.plotting import plot

## (a) Find the plane's ground speed in terms of $v_0, V, \phi$ (the angle by which the plane heads to the north of east), and the plane's position. 

In [96]:
var('v0 V y phi') # create variables

(v0, V, y, phi)

In [97]:
v_vec = v0 + V*y # superposition of initial velocity and velocity from wind

In [98]:
# break velocity in to components 
v_vec_x = v0*cos(phi) + V*y 
v_vec_y = v0*sin(phi) 
v_vec_z = 0

In [99]:
groundspeed = sqrt(v_vec_x**2 + v_vec_y**2 + v_vec_z**2)
groundspeed.simplify()

sqrt(V**2*y**2 + 2*V*v0*y*cos(phi) + v0**2)

## (b) Write down the time of flight as an integral of the form 
$$L = \int\limits_{0}^{D} f\, dx$$
Show that if we assume that $y'$ and $\phi$ both remain small, then the integrand $f$ takes the approximate form $$f = \frac{1+0.5y'^2}{1+ky}$$ where $k = V/v_0$

now time integral can be written as 

In [100]:
var('dx dphi dy yprime k yprimeprime l x D') # create variables

(dx, dphi, dy, yprime, k, yprimeprime, l, x, D)

In [101]:
ds = sqrt(dx**2+dy**2)

We can pull a $dx^2$ out of the expression to get 

In [102]:
ds = sqrt(1+(dy/dx)**2)*dx
ds

dx*sqrt(1 + dy**2/dx**2)

Which can be written as 

In [103]:
ds = dx*sqrt(1+yprime**2)
ds

dx*sqrt(yprime**2 + 1)

In [104]:
integrand2 = ds/v_vec

So the time integral becomes $$L = \int\limits_{0}^{D} \frac{ds}{v}\, dx$$
$$L = \int\limits_{0}^{D} \frac{\sqrt(y'^2+1)}{v_0+Vy}\, dx$$

Using a binomial approximation for small $y'$ and $k = V/v_0$, we get 
$$L = \int\limits_{0}^{D} \frac{1+ 0.5y'^2}{1+ky}\, dx$$

In [105]:
f = (1+0.5*yprime**2)/(v0 + V*y)
f

(0.5*yprime**2 + 1)/(V*y + v0)

## (c) Write down the E-L equation that determines the best path. To solve it, guess that $y(x) = \lambda x(D-x)$. Show that it satisfies the E-L equation, given $\lambda = \frac{\sqrt(4+2k^2D^2)-2}{kD^2}$. How far north does this path take the plane, if $ D = 2000$ miles, $v_0 = 500$ mph, and the wind shear is $V = 0.5$ mph/mi? How much time does the plane save by following this path?

E-L equation tells us 
$$\frac{∂f}{∂y} = \frac{d}{dx}\frac{\partial f}{\partial y'}$$

We have the integrand $f$ and can calculate the LHS

In [106]:
LHS = diff(f, y)
LHS

-V*(0.5*yprime**2 + 1)/(V*y + v0)**2

In [107]:
partialyprime = diff(f, yprime)
partialyprime

1.0*yprime/(V*y + v0)

In [108]:
RHS = (yprimeprime*(v0+V*y) - V*yprime**2)/(v0+V*y)**2
RHS

(-V*yprime**2 + yprimeprime*(V*y + v0))/(V*y + v0)**2

So full E-L expression is the below expression set equal to $0$

In [109]:
EL = LHS - RHS
EL.simplify()

(V*yprime**2 - V*(0.5*yprime**2 + 1) - yprimeprime*(V*y + v0))/(V*y + v0)**2

The solution to this given by the text is $y = \lambda x(D-x)$

for $\lambda = $

In [110]:
#l = (sqrt(4+2*(V/v0)**2*D**2)-2)/((V/v0)*D**2)
#l

In [111]:
soln = l*x*(D-x)
soln

l*x*(D - x)

In [117]:
prime = diff(soln, x)
prime.simplify()

l*(D - 2*x)

In [118]:
primeprime = diff(prime, x)
primeprime.simplify()

-2*l

In [135]:
num = (V*prime**2-V*(0.5*prime**2+1)-primeprime*(V*soln+v0))/(V*soln+v0)**2
num.simplify()

(V*l**2*(-D + 2*x)**2 - V*(0.5*l**2*(-D + 2*x)**2 + 1) + 2*l*(V*l*x*(D - x) + v0))/(V*l*x*(D - x) + v0)**2

We know that this result is $0$ if $l =$

In [146]:
solve(num, l) # solve expression for lambda

[(-2.0*v0 + 2.0*sqrt(0.5*D**2*V**2 + v0**2))/(D**2*V),
 -(2.0*v0 + 2.0*sqrt(0.5*D**2*V**2 + v0**2))/(D**2*V)]

In [145]:
lambdaa = (-2.0*v0 + 2.0*sqrt(0.5*D**2*V**2 + v0**2))/(D**2*V) # choose positive root 

We can now apply the numbers given to us in the problem to get specifics. To get the maximum north displacement, we can solve the solution expression for $y' = 0$


In [144]:
solve(prime, x)

[D/2]

So the maximum noth displacement happens for $x = D/2$

We can use this to find lambda for the values given, and from there find $y(D/2)$

In [149]:
maxl = lambdaa.subs({D:2000, v0: 500, V:0.5, x:2000/2}) # plug in given values 
maxl

0.000366025403784439

In [150]:
soln.subs({l:maxl, x:2000/2, D:2000})

366.025403784439

So max north displacement is $366$ miles. To find the time saved, we can calulate the actual time integral $L$. Without following the path, total time taken is $D/v0 = 2000/500 = 4$ hours. Calculations for the amount of time required when following optimal path:

In [161]:
integrand3 = (1+0.5*prime**2)/(v0 + V*soln)
integrand3

(0.5*(-l*x + l*(D - x))**2 + 1)/(V*l*x*(D - x) + v0)

In [165]:
evaluate = integrand3.subs({D:2000, v0: 500, V:0.5, l:maxl})
evaluate

(0.267949192431123*(1 - 0.001*x)**2 + 1)/(0.000183012701892219*x*(2000 - x) + 500)

In [166]:
integrate(evaluate, (x, 0, 2000))

3.55577268553852

Following the optimal path takes $3.56$ hours, and so saves $0.44$ hours of travel time. 